R/iStar.R

```diff
@@ -275,48 +275,50 @@ iStarPostTrade <- function(MktData
     }
     
     secMktDataDaily <- secMktData[endpoints(secMktData, 'days')]
-    secMktDataDaily[, 'MktQty']   <- period.apply(secMktData[, 'MktQty'], endpoints(secMktData[, 'MktQty'], 'days'), sum)
+    secMktDataDaily[, 'MktQty'] <- period.apply(secMktData[, 'MktQty'], endpoints(secMktData[, 'MktQty'], 'days'), sum)
     secMktDataDaily$MktValue <- secMktDataDaily[, 'MktPrice'] * secMktDataDaily[, 'MktQty']
-
+    
+    cat(names(MktData)[s], "(days =", paste0(nrow(secMktDataDaily), "):"), "\n")
+    
     # Arrival Price
+    arrPriceIdx <- endpoints(secMktData, 'days') + 1L
+    arrPriceIdx <- arrPriceIdx[-length(arrPriceIdx)]
     if (any(colnames(secMktDataDaily) == 'Bid') & any(colnames(secMktDataDaily) == 'Ask')) {# first bid-ask spreads midpoint
-      arrPrice <- 0.5 * (secMktDataDaily[1, 'Ask'] + secMktDataDaily[1, 'Bid'])
+      arrPrice <- 0.5 * (secMktData[arrPriceIdx, 'Ask'] + secMktData[arrPriceIdx, 'Bid'])
     } else {# proxy
-      arrPrice <- secMktDataDaily[1, 'MktPrice']
+      arrPrice <- secMktData[arrPriceIdx, 'MktPrice']
     }
     
-    cat(names(MktData)[s], "(days =", paste0(nrow(secMktDataDaily), "):"), "\n")
-    
-    for (t in 1:(nrow(secMktDataDaily) - horizon + 1)) {
+    for (t in 1:(nrow(secMktDataDaily) - horizon + 1L)) {
       # Rolling periods and dates (with consistent timestamps, if needed)
-      hStop <- t + horizon - 1
-      refIdx <- hStop + 1
-      periodIdxs[[s]] <- (horizon + 1):(nrow(secMktDataDaily) + 1)
-      nextDayLastDate[[s]] <- as.Date(last(index(secMktDataDaily))) + 1 # last "next day" may be a non-business day!
+      hStop <- t + horizon - 1L
+      # refIdx <- hStop + 1L
+      periodIdxs[[s]] <- (horizon + 1L):(nrow(secMktDataDaily) + 1L)
+      nextDayLastDate[[s]] <- as.Date(last(index(secMktDataDaily))) + 1L # last "next day" may be a non-business day!
       nextDayDates[[s]] <- c(as.Date(index(secMktDataDaily)[periodIdxs[[s]][1:(nrow(secMktDataDaily) - horizon)]]), nextDayLastDate[[s]])
       
       # Volatility (on close-to-close prices, annualized)
       secCloseReturns <- Return.calculate(secMktDataDaily[t:hStop, 'MktPrice'], 'log')
-      secAnnualVol[refIdx, s] <- as.numeric(sd.annualized(secCloseReturns, scale = yrBizdays))
+      secAnnualVol[hStop, s] <- as.numeric(sd.annualized(secCloseReturns, scale = yrBizdays))
       
       # Average Market Volume
-      ADV[refIdx, s] <- mean(sum(secMktDataDaily[t:hStop, 'MktQty']))
+      ADV[hStop, s] <- mean(sum(secMktDataDaily[t:hStop, 'MktQty']))
```

Comment:
We can remove sum here, as we are computing mean of a scalar which is the scalar, and the scalar is the sum of total volume for each day, which means ADV is significantly overstated

Reply:
Oh, yes, vectorization.